### PR TITLE
First merge of NextGen design transformations

### PR DESCRIPTION
**Pull request by pnikiel**

### CODE REVIEW

File: AddressSpace/templates/designToClassBody.jinja [hunk 1]
```diff
@@ -0,0 +1,708 @@
+{# © Copyright CERN, 2015.                                                       #}
```
[the remark this answers is not in the document]
Comment (ben-farnham) [reply AddressSpace/templates/designToClassBody.jinja:1]: Going to make a vague comment here, more of a 'feeling' than something I can point to as a concrete example...

Are we making good usage of C++' s polymorphism? Compile time polymorphism (templates) and runtime polymorphism (inheritance etc). The jinja templates are _waaay_ better than XSLT, but still look undoubtedly scary. Any reduction in template complexity is a benefit.

Put another way - are we using jinja templates to brute force generate our way out of the requirements of different types; might we also be able to create (static - i.e. generic to _all_ servers) C++ template functions then call those C++ template functions (from the jinja template) to further reduce the scariness ?

[the remark this answers is not in the document]
Comment (ben-farnham) [reply AddressSpace/templates/designToClassBody.jinja:1]: Another thing that may help reduce jinja template complexity is moving parsing Design.XML into semantic logic fully into DesignInspector.py. I struggled with turning the design.xml of cachevariable/configentry configRestrcition elements into C++(/XSD) code in the templates.

In the end I pushed entirely the parsing etc logic into DesignInspector and basically return a dict of strings into the template. I think it helped: the template has no lingering Design.xml 'smell' about it: it only deals with what DesignInspector gives it back.

See what you think...(have a [look here](https://github.com/quasar-team/quasar/blob/OPCUA-1746_BF_NextGen_design_transformations/FrameworkInternals/DesignInspector.py)) function 'parseRestriction' and have a look where it's called in the Configuration jinja templates

File: FrameworkInternals/transformDesign.py [hunk 2]
```diff
@@ -71,42 +76,84 @@ class FieldIds(enum.Enum):
...
+    if not os.path.isdir(outputDirectory):
+        os.makedirs(outputDirectory)
+    designInspector = DesignInspector(designXmlPath)
+    transformDir = os.path.dirname(transformPath)
+    commonTemplatesLoader = jinja2.FileSystemLoader(os.path.join(transformDir, '..', '..', 'Common', 'templates'))
+    moduleTemplatesLoader = jinja2.FileSystemLoader(transformDir)
+    env = jinja2.Environment(loader=jinja2.ChoiceLoader([commonTemplatesLoader, moduleTemplatesLoader]))
+    env.filters['capFirst']  = capFirst
+    env.filters['debug'] = templateDebug
+    env.filters['cppCommentsToCmakeComments'] = cppCommentsToCmakeComments
```
[the remark this answers is not in the document]
Comment (pnikiel) [reply FrameworkInternals/transformDesign.py:138]: It's a filter which transforms blocks of comments in C++ syntax into CMake comments.

Without this we would need to define header contents (that longish text about copyrights etc) in every possible comment style - CMake comments (#), C++ comments (/**), HTML output (<!--) and who knows what else.

Instead I propose to store only the C++ version of the header and use a filter to derive the CMake or HTML version or so. Redundancy reduction. What do you think?

File: FrameworkInternals/transformDesign.py [hunk 3]
```diff
@@ -71,42 +76,84 @@ class FieldIds(enum.Enum):
...
+    """ additionalParam - a dictionary that will be passed to the transform """
+    outputDirectory = os.path.dirname(outputFile)
+    if not os.path.isdir(outputDirectory):
+        os.makedirs(outputDirectory)
+    designInspector = DesignInspector(designXmlPath)
+    transformDir = os.path.dirname(transformPath)
+    commonTemplatesLoader = jinja2.FileSystemLoader(os.path.join(transformDir, '..', '..', 'Common', 'templates'))
+    moduleTemplatesLoader = jinja2.FileSystemLoader(transformDir)
+    env = jinja2.Environment(loader=jinja2.ChoiceLoader([commonTemplatesLoader, moduleTemplatesLoader]))
+    env.filters['capFirst']  = capFirst
```
[the remark this answers is not in the document]
Comment (pnikiel) [reply FrameworkInternals/transformDesign.py:136]: I think it's a very good idea. Will do.

File: FrameworkInternals/transformDesign.py [hunk 4]
```diff
@@ -71,42 +76,84 @@ class FieldIds(enum.Enum):
...
+def transformDesignByJinja(designXmlPath, transformPath, outputFile, additionalParam):
+    """ additionalParam - a dictionary that will be passed to the transform """
+    outputDirectory = os.path.dirname(outputFile)
+    if not os.path.isdir(outputDirectory):
+        os.makedirs(outputDirectory)
+    designInspector = DesignInspector(designXmlPath)
+    transformDir = os.path.dirname(transformPath)
+    commonTemplatesLoader = jinja2.FileSystemLoader(os.path.join(transformDir, '..', '..', 'Common', 'templates'))
+    moduleTemplatesLoader = jinja2.FileSystemLoader(transformDir)
+    env = jinja2.Environment(loader=jinja2.ChoiceLoader([commonTemplatesLoader, moduleTemplatesLoader]))
```
[the remark this answers is not in the document]
Comment (pnikiel) [reply FrameworkInternals/transformDesign.py:135]: The problem I was solving here was that jinja sources are stored in possibly multiple dirs, for example for a transform in Device module:
* in templates/, e.g. designToDeviceHeader
* in common templates, for example for common headers etc.
FileSystemLoader unfortunately does not permit to supply multiple search directories, so we achieve it by ChoiceLoader: if the transform is found via the 1st FileSystemLoader, it's taken, if not, it is looked for in the 2nd FileSystemLoader.

Makes sense?

File: AddressSpace/templates/designToClassBody.jinja (hunk 1 above)
[the remark this answers is not in the document]
Comment (pnikiel) [reply AddressSpace/templates/designToClassBody.jinja:1]: For the 2nd comment - I did use Jinja macros for instance for copyright header placement.
I believe that there are 'common blocks' which I implemented rather as Oracle functions (example: https://github.com/quasar-team/quasar/blob/OPCUA-1746_NextGen_design_transformations/FrameworkInternals/Oracle.py#L150 )
Certainly it was possible to implement this as Jinja macro but expressing such logic in Python was much more obvious thus I tool the latter way.

For the 3rd comment - there were places where I already moved from XSLT/Jinja2 into C++, for example look at the usage of m_effectiveParentNodeId. This is not about templates or runtime polymorphism though. I probably do not understand your specific idea. 
Note that quasar API was supposed to stay intact in this rework. But still, if you have a specific improvement why not remarking on  it in the next step, after we're finished with this.

For the 4th comment - I didn't yet review your idea. Did you identify places in desginToClassBody where this approach would be attractive?

**Pull request closed without merging**